flac.go

```diff
@@ -327,60 +314,44 @@ func (stream *Stream) Seek(sampleNum int64, whence int) (result int64, err error
 
 	rs := stream.r.(io.ReadSeeker)
 
-	var point meta.SeekPoint
-	switch whence {
-	case io.SeekStart:
-		point = stream.searchFromStart(sampleNum)
-	case io.SeekCurrent:
-		point, err = stream.searchFromCurrent(sampleNum, rs)
-	case io.SeekEnd:
-		point = stream.searchFromEnd(sampleNum)
-	default:
-		return 0, ErrInvalidSeek
+	isBiggerThanStream := stream.Info.NSamples != 0 && stream.Info.NSamples < sampleNum
+	if isBiggerThanStream || sampleNum < 0 {
+		return 0, fmt.Errorf("unable to seek to sample number %d", sampleNum)
 	}
+	point := stream.searchFromStart(sampleNum)
 
+	_, err := rs.Seek(stream.dataStart+int64(point.Offset), io.SeekStart)
 	if err != nil {
 		return 0, err
 	}
-
-	_, err = rs.Seek(stream.dataStart+int64(point.Offset), io.SeekStart)
-	return int64(point.SampleNum), err
-}
-
-func (stream *Stream) searchFromCurrent(sample int64, rs io.ReadSeeker) (p meta.SeekPoint, err error) {
-	o, err := rs.Seek(0, io.SeekCurrent)
-	if err != nil {
-		return p, err
-	}
-
-	offset := o - stream.dataStart
-	for _, p = range stream.seekTable.Points {
-		if int64(p.Offset) >= offset {
-			return stream.searchFromStart(int64(p.SampleNum) + sample), nil
+	for {
+		offset, err := rs.Seek(0, io.SeekCurrent)
+		if err != nil {
+			return 0, err
+		}
+		frame, err := stream.ParseNext()
+		if err != nil {
+			return 0, err
+		}
+		if frame.SampleNumber()+uint64(frame.BlockSize) >= sampleNum {
+			_, err := rs.Seek(offset, io.SeekStart)
+			return frame.SampleNumber(), err
 		}
 	}
-	return p, nil
-}
-
-// searchFromEnd expects sample to be negative.
-// If it is positive, it's ok, the last seek point will be returned.
-func (stream *Stream) searchFromEnd(sample int64) (p meta.SeekPoint) {
-	return stream.searchFromStart(int64(stream.Info.NSamples) + sample)
 }
 
-func (stream *Stream) searchFromStart(sample int64) (p meta.SeekPoint) {
-	var last meta.SeekPoint
-	var i int
-	for i, p = range stream.seekTable.Points {
-		if int64(p.SampleNum) >= sample {
-			if i == 0 {
-				return p
-			}
-			return last
+func (stream *Stream) searchFromStart(sample uint64) meta.SeekPoint {
+	if len(stream.seekTable.Points) == 0 {
+		panic("do not reach this lol")
+	}
+	prev := stream.seekTable.Points[0]
+	for _, p := range stream.seekTable.Points {
+		if p.SampleNum+uint64(p.NSamples) >= sample {
+			return prev
 		}
-		last = p
+		prev = p
 	}
-	return p
+	panic("unreachable")
 }
 
 func (stream *Stream) makeSeekTable() (err error) {
```

Comment:
Add a short doc comment. Something along the lines of:

```go
// makeSeekTable creates a seek table with seek points to each frame of the FLAC stream.
```

We may consider making this configurable in the future. For now, the memory use of the seek table was small enough and the CPU time taken to construct the seek table large enough to justify storing seek points to each frame.